systemd/systemd.go

```diff
@@ -153,30 +152,81 @@ func (*systemd) LogReader(serviceNames []string, n string, follow bool) (io.Read
 	return JournalctlCmd(serviceNames, n, follow)
 }
 
-var statusregex = regexp.MustCompile(`(?m)^(?:(.*?)=(.*))?$`)
+var statusregex = regexp.MustCompile(`(?m)^(?:(.+?)=(.*)|(.*))?$`)
 
-func (s *systemd) Status(serviceNames ...string) ([]*client.ServiceInfo, error) {
+type ServiceStatus struct {
+	Daemon          string
+	ServiceFileName string
+	Enabled         bool
+	Active          bool
+}
+
+func (s *systemd) Status(serviceNames ...string) ([]*ServiceStatus, error) {
 	cmd := make([]string, len(serviceNames)+2)
-	copy(cmd[2:], serviceNames)
 	cmd[0] = "show"
 	cmd[1] = "--property=Id,ActiveState,UnitFileState"
+	copy(cmd[2:], serviceNames)
 	bs, err := SystemctlCmd(cmd...)
 	if err != nil {
 		return nil, err
 	}
 
-	sts := make([]*client.ServiceInfo, 0, len(serviceNames))
-	cur := &client.ServiceInfo{}
+	sts := make([]*ServiceStatus, 0, len(serviceNames))
+	cur := &ServiceStatus{}
+	expected := map[string]bool{
+		"Id":            true,
+		"Type":          true,
+		"ActiveState":   true,
+		"UnitFileState": true,
+	}
 
 	for _, bs := range statusregex.FindAllSubmatch(bs, -1) {
 		if len(bs[0]) == 0 {
 			// systemctl separates data pertaining to particular services by an empty line
+			missing := make([]string, 0, len(expected))
+			for k, v := range expected {
+				if v {
+					missing = append(missing, k)
+				}
+			}
+			if len(missing) > 0 {
+				return nil, fmt.Errorf("cannot get service status: missing %s in ‘systemctl show’ output", strings.Join(missing, ", "))
+
+			}
 			sts = append(sts, cur)
-			cur = &client.ServiceInfo{}
+			if len(sts) > len(serviceNames) {
+				break // wut
+			}
+			if cur.ServiceFileName != serviceNames[len(sts)-1] {
+				return nil, fmt.Errorf("cannot get service status: queried status of %q but got status of %q", serviceNames[len(sts)], cur.ServiceFileName)
+			}
+
+			cur = &ServiceStatus{}
+			expected = map[string]bool{
+				"Id":            true,
+				"Type":          true,
+				"ActiveState":   true,
+				"UnitFileState": true,
+			}
 			continue
 		}
+		if len(bs[3]) > 0 {
+			return nil, fmt.Errorf("cannot get service status: bad line %q in ‘systemctl show’ output", bs[3])
+		}
 		k := string(bs[1])
 		v := string(bs[2])
+		if b, ok := expected[k]; b {
+			expected[k] = false
```

Comment:
I would keep expected as an array, and have seen instead of doing this and the resetting of expected, would seem more readable to me